rest/route/github.go

```diff
@@ -306,6 +319,135 @@ func (gh *githubHookApi) AddIntentForPR(pr *github.PullRequest) error {
 	return nil
 }
 
+func (gh *githubHookApi) createVersionForTag(ctx context.Context, event *github.PushEvent) error {
+	if err := validatePushTagEvent(event); err != nil {
+		return err
+	}
+
+	pusherID := int(event.GetPusher().GetID())
+	u, err := gh.sc.FindUserByGithubUID(pusherID)
+	if err != nil {
+		return errors.Errorf("Error finding evergreen user for Github UID %d", pusherID)
+	}
+	if u == nil {
+		return errors.Errorf("No user found with Github UID %d", pusherID)
+	}
+	// TODO: add validation for user
+	tag := model.GitTag{
+		Tag:             strings.TrimPrefix(event.GetRef(), refTags),
+		Pusher:          event.GetPusher().GetLogin(),
+		PusherGithubUID: pusherID,
+	}
+	ownerAndRepo := strings.Split(event.Repo.GetFullName(), "/")
+	hash := event.HeadCommit.GetSHA()
+
+	projectRefs, err := gh.sc.FindEnabledProjectRefsByOwnerAndRepo(ownerAndRepo[0], ownerAndRepo[1])
+	if err != nil {
+		return errors.Wrapf(err, "error finding projects of repo '%s'", ownerAndRepo)
+	}
+	if len(projectRefs) == 0 {
+		return errors.Wrapf(err, "no projects found for repo '%s'", ownerAndRepo)
+	}
+	catcher := grip.NewBasicCatcher()
+	for _, pRef := range projectRefs {
+		// if a version for this revision exists for this project, add tag
+		existingVersion, err := gh.sc.FindVersionByProjectAndRevision(pRef.Identifier, hash)
```

Comment:
When would this happen? (I'm actually wondering since I don't know the workflow)

Reply:
So if you're pushing a tag to a commit, it means this commit has already been pushed (may need to experiment with what happens when you push a tag to a commit in a PR and then merge the PR but this is the case here anyway), which means it already had a version created for it on the waterfall (assuming this project was activated at the time this commit was pushed, or else we move on). So this query finds that version on the waterfall and adds a git tag to it so we can start displaying git tags on the waterfall.

We look at every valid project because there might be projects for other branches that also have this commit, and we want to make sure we handle any that is applicable.

Reply:
Yea that was one of the scenarios I was thinking of. Specifically, can you have a commit locally (not pushed yet), annotate that commit with a tag, then push your changes? If so what do we want to do in that scenario?

Reply:
Okay good question. If I push my changes straight to upstream, I still have to push the tag separately. If the version hasn't been created yet I guess we won't trigger a version for the tag, or add the tag to the version.

If a tag was pushed to a commit that was part of a code review, this won't trigger the webhook (this seems to me like the correct behavior). When the work to add tags to the waterfall is being done then, this may also involve some additional work to check if commits have tags, but that would be for visibility and not triggered behavior.

When we write documentation for this process we can specify that tags _should_ be pushed for revisions that have already built on the waterfall (this also does make maybe the most sense).

Reply:
I agree with Annie. This is also how the scope document is written. The first goal is to "Process git tag webhook to trigger a release build."